engine/src/main/java/org/camunda/bpm/engine/impl/bpmn/behavior/ParallelGatewayActivityBehavior.java

```diff
@@ -63,18 +66,52 @@ public void execute(ActivityExecution execution) throws Exception {
     lockConcurrentRoot(execution);
     
     List<ActivityExecution> joinedExecutions = execution.findInactiveConcurrentExecutions(activity);
-    int nbrOfExecutionsToJoin = execution.getActivity().getIncomingTransitions().size();
-    int nbrOfExecutionsJoined = joinedExecutions.size();
+    
+    List<ActivityExecution> finalExecutions = new ArrayList<ActivityExecution>();
+    
+    Map<String, List<ActivityExecution>> mappedExecutions = mapExecutionsToTransitions(joinedExecutions, finalExecutions);
+    
+    List<PvmTransition> incomingTransitions = execution.getActivity().getIncomingTransitions();
+    for (PvmTransition incomingTransition : incomingTransitions) {
+      List<ActivityExecution> sequenceExecutions = mappedExecutions.get(incomingTransition.getId());
+      if (sequenceExecutions == null || sequenceExecutions.isEmpty()){
+        //TODO if required for logging we can go through all of them?
+        break;
```

Comment:
Does this work in case some executions have a transition id set and some have not?

Assuming the following:
A parallel join with three incoming sequence flows, a, b, c.
Three executions: One with transition ID == null, one with transitiion ID == b, one with transition ID == c.

Let's further assume that the first incomingTransition in this loop is transition a. The loop breaks and finalExecutions contains only the first execution. However, the gateway could fire because there are executions for transitions b and c.

One way of fixing could be to continue in the loop where you break now and to break if finalExecutions has reached the desired size. You can also write a unit test for this case. To set the transition ID to null, you could write a custom command in the test that manipulates the field on ExecutionEntity.

I think, such a case can occur when people upgrade Camunda with running process instances and where execution 1 has reached the parallel gateway before the upgrade and the others after the upgrade.